File: app/models/item_wiki.py
# -*- coding: utf-8 -*-
"""
数据模型: 物品Wiki - 物品的类定义
"""

import uuid
import logging
from datetime import datetime
from typing import Optional
from sqlalchemy import Column, String, Text, DateTime, Integer
from app.models import Base


class ItemWiki(Base):
    """
    物品Wiki模型 - 物品的类定义

    存储物品的通用属性，作为物品库存记录的模板。
    每个库存记录(Item)关联到一个Wiki条目。
    """
    __tablename__ = 'item_wikis'

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))

    name = Column(String(100), nullable=False, index=True)
    description = Column(Text, nullable=True)

    default_unit = Column(String(20), nullable=True)  # 默认单位：个、盒、瓶等

    suggested_expiry_days = Column(Integer, nullable=True)  # 建议保质期（天）

    storage_location = Column(String(100), nullable=True)  # 建议存放位置：冷藏、冷冻、常温

    notes = Column(Text, nullable=True)  # 备注信息

    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)

    def __repr__(self):
        return f"<ItemWiki(id='{self.id}', name='{self.name}')>"

    @property
    def inventory_count(self) -> int:
        """获取该Wiki下的库存记录数量"""
        try:
            from app.services.database import db_service
            from app.models.item import ItemStatus

            with db_service.session_scope() as session:
                count = session.query(Item).filter(
                    Item.wiki_id == self.id,
                    Item.status != ItemStatus.CONSUMED
                ).count()
                return count
        except Exception as e:
            logging.error(f"获取Wiki库存数量失败: {e}")
            return 0


class ItemWikiCategory(Base):
    """
    物品Wiki分类模型
    """
    __tablename__ = 'item_wiki_categories'

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))

    name = Column(String(50), nullable=False, unique=True, index=True)

    icon = Column(String(50), nullable=True)  # 图标名称

    color = Column(String(20), nullable=True)  # 颜色（十六进制）

    sort_order = Column(Integer, nullable=False, default=0)  # 排序顺序

    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    def __repr__(self):
        return f"<ItemWikiCategory(id='{self.id}', name='{self.name}')>"

File: mobile/lib/models/item_wiki.dart
class ItemWiki {
  const ItemWiki({
    required this.id,
    required this.name,
    this.icon,
    this.description,
    this.categoryId,
    this.categoryName,
    this.defaultUnit,
    this.suggestedExpiryDays,
    required this.defaultReminderDays,
    this.storageLocation,
    this.notes,
    this.imagePath,
    required this.createdAt,
    required this.updatedAt,
    this.inventoryCount = 0,
  });

  final String id;
  final String name;
  final String? icon;
  final String? description;
  final String? categoryId;
  final String? categoryName;
  final String? defaultUnit;
  final int? suggestedExpiryDays;
  final int defaultReminderDays;
  final String? storageLocation;
  final String? notes;
  final String? imagePath;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int inventoryCount;

  factory ItemWiki.fromMap(Map<String, Object?> map) {
    return ItemWiki(
      id: map['id'] as String,
      name: map['name'] as String,
      icon: map['icon'] as String?,
      description: map['description'] as String?,
      categoryId: map['category_id'] as String?,
      categoryName: map['category_name'] as String?,
      defaultUnit: map['default_unit'] as String?,
      suggestedExpiryDays: map['suggested_expiry_days'] as int?,
      defaultReminderDays: (map['default_reminder_days'] as int?) ?? 3,
      storageLocation: map['storage_location'] as String?,
      notes: map['notes'] as String?,
      imagePath: map['image_path'] as String?,
      createdAt: DateTime.parse(map['created_at'] as String),
      updatedAt: DateTime.parse(map['updated_at'] as String),
      inventoryCount: (map['inventory_count'] as int?) ?? 0,
    );
  }

  Map<String, Object?> toMap() {
    return {
      'id': id,
      'name': name,
      'icon': icon,
      'description': description,
      'category_id': categoryId,
      'default_unit': defaultUnit,
      'suggested_expiry_days': suggestedExpiryDays,
      'default_reminder_days': defaultReminderDays,
      'storage_location': storageLocation,
      'notes': notes,
      'image_path': imagePath,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}
